[JIT] Add support for default args in class types

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42988 [JIT] Add support for default args in class types**

**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

Differential Revision: [D23762013](https://our.internmc.facebook.com/intern/diff/D23762013)

Commit:
[JIT] Add support for default args in class types

**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

gmagogsfm commented on test/jit/test_class_type.py:
```diff
@@ -1013,6 +1013,79 @@ def foo():
             y.my_list = new_list
             return y
 
+    def test_default_args(self):
```
This test case doesn't seem to exercise the logic of padding default values (in case of a fn only has part of its arguments having default values). Could you add one?

SplitInfinity replied:
Do you mean something like:
```
        class ClassWithDefaultArgs:
            def __init__(
                self,
                a: int,
                b: List[int] = [1, 2, 3],
            ):
```

gmagogsfm replied:
Yep. 

On a related note, maybe we can have a test covering the exception thrown with mutable default values as well?

SplitInfinity commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -217,6 +217,33 @@ TypePtr ScriptTypeParser::parseType(const std::string& str) {
 }
 
+bool checkMutableFunctionDefault(const IValue& def) {
+  if (def.isList() && def.isGenericDict()) {
```
Yes oops :'(

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -217,6 +217,33 @@ TypePtr ScriptTypeParser::parseType(const std::string& str) {
 }
 
+bool checkMutableFunctionDefault(const IValue& def) {
+  if (def.isList() || def.isGenericDict()) {
```
can we invoke the existing logic here instead of copy-pasta ? 

https://github.com/pytorch/pytorch/blob/fcc10d75e1a0c1c2474de559124620ca6b580313/torch/csrc/jit/python/script_init.cpp#L223

eellison reviewed (commented):
didnt do full review, just comment here

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -57,7 +57,7 @@ def _is_new_style_class(cls):
 
 def _compile_and_register_class(obj, rcb, qualified_name):
-    ast = get_jit_class_def(obj, obj.__name__)
+    ast = get_jit_class_def(obj, obj.__name__, parse_defaults=True)
```
Good question, but parsing defaults all the time breaks a lot of things because this same code path is also used to compile module interfaces. Parsing defaults in that breaks a bunch of module interface tests and I considered it out of scope for this PR.

eellison replied:
Why do the tests fail?

SplitInfinity replied:
I've noticed two categories of failures:
1) The function or interface has a default argument that gets inferred to be a different type when parsed as an AST constant. For example, the function definition has argument like `lambda=0` where `lambda` is not annotated or if it is, it's below the `def` as a mypy style comment. So, JIT thinks it's a int when it is actually a float. These cases can be fixed by replacing the literals with `0.0` or `0e0`, etc.
2) The function or interface has a mutable default value for an argument, like a list or a dict. This can be fixed by assigning those to a variable and then using that variable instead, or by using `None` and then assigning the actual default value in the body. One of the issues with the former solution (assigning the defaults to a variable in the containing scope and then using that variable instead) is that they are resolved using `toSugaredValue` to a generic `PythonValue` instance that you can't use as a value when processing the assignment.

eellison replied:
I think those two failures are not to hard to fix, and should be in scope of this PR instead of introducing a code split

SplitInfinity commented on test/jit/test_class_type.py:
```diff
@@ -1013,6 +1013,79 @@ def foo():
+            return a.get_int() + a.get_list()[2] + a.get_tup()[1]
+
+        def some_defaults() -> int:
+            a: ClassWithDefaultArgs = ClassWithDefaultArgs(b=[5, 6, 7])
```
Noted, will change the object variable name to `obj`

eellison reviewed (commented):
What do you think about the PR here ? https://github.com/pytorch/pytorch/pull/31344/files

SplitInfinity commented:
> What do you think about the PR here ? https://github.com/pytorch/pytorch/pull/31344/files

I think it's more or less what I have here. However, one thing I don't see there is checks for mutable function defaults, which, judging by the failures on this PR, happen quite a bit.

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

eellison commented on torch/functional.py:
```diff
@@ -1158,12 +1158,12 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
 
     @overload  # noqa: 749
     def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa: 749
-        # type: (Tensor, Optional[number], Optional[List[int]], bool, Optional[Tensor], Optional[int]) -> Tensor
```
Why are there any changes in this file ?

SplitInfinity replied:
These functions are getting pulled in and scripted due to recursive processing in some unit tests. This was part of the reason I added that `parse_defaults` flag and set it to `True` only for classes.

eellison commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -217,13 +217,81 @@ TypePtr ScriptTypeParser::parseType(const std::string& str) {
 }
 
+bool checkMutableFunctionDefault(const IValue& def) {
+  // TODO: Renable.
```
I dont think we should disable this to land the PR - can you comment why this was necessary?

SplitInfinity replied:
That was not my intention - I'm still trying to wade through a bunch of broken tests to figure out what's going wrong with trying to parse default arguments everywhere (not just in class types). I have not yet requested review for this reason.

SplitInfinity replied:
I want to reenable this, but in the process of debugging some broken tests, I found some that somehow do have list-literals as default arguments, which are considered mutable. I was not able to find them in the codebase anywhere, so I think they must be generated somehow (perhaps with `native_functions.yaml`). I was going to bring this up after sifting through all of the tests to figure out why parsing defaults everywhere causes so many problems.

SplitInfinity commented:
I'm still wading through a huge pool of failing tests to figure out what exactly is going wrong with parsing defaults in all cases (with regular functions, interfaces, module methods and class type methods). I only run JIT tests locally so I'm relying on CI to tell me what else is wrong.

It is for this reason that I have not yet re-requested review despite pushing many updates in the last few days.

eellison commented:
@SplitInfinity ah gotcha, thought you rerequested. Ignore me then :).

SplitInfinity commented:
@eellison @gmagogsfm  

Okay, now that tests are mostly passing, I think we can talk about all of the problems I encountered, what the fixes should be, and whether it's worth continuing with this PR.

### We cannot use variables from the containing scope as default values
For reasons that I do not completely understand, the resolution callback we create to resolve names from the containing scope cannot resolve variables from the containing scope that are used as default values for arguments of a function. They are simply absent from the data structures that we use to create those resolution callbacks (e.g. `fn.__code__.co_freevars`). So, if the function is declared like so:
```
my_var = [0, 0]
def my_function(a: List[int] = my_var): ...
```
we cannot resolve `my_var`. However, something like this is okay:
```
my_var = [0, 0]
def my_function(a: Optional[List[int]] = None):
  if a is None:
    a = my_var
```
because `my_var` is used in the function body and can be resolved by the resolution callback we create. But this might require a lot of downstream changes.

### There are kernels that are being recursively compiled that have mutable default argument values
While wading through broken tests, I found that some of them were failing because they use a `maxpool_2d` kernel that has a list literal as the default value of an argument. This is not allowed because mutations to the argument will persist across calls. However, I was unable to find this kernel; my best guess is that it is generated from `native_functions.yaml`:
```
- func: max_pool2d(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1, bool ceil_mode=False) -> Tensor
```
If this kernel was committed in source code form, the fix for this would be similar to the one above: change the type of the argument to `Optional` and conditionally initialize it with the default value in the body. I disabled the check for mutable default arguments due to these kernels but I would like to reenable it before landing.

### There are kernels that are being recursively compiled whose arguments cannot be expressed well in the JIT type system
`norm` in `torch/functional.py` has an argument that should be `Union[str, int, float]`. This cannot be expressed in TorchScript, so I had to change the type of that argument to `Any` and modify the source code to do a bunch of type refinement to get things to work.

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -190,6 +190,7 @@ def _forward(self):
             In this case, the `__name__` attribute of the function object is "_forward",
             but we want the result AST to have the name "forward".
         self_name: If this function is a method, what the type name of `self` is.
+        parse_defaults: If True, the default values for functions are parsed in included in the result AST.
```
in included in -> included in

SplitInfinity replied:
Removed altogether since it's no longer used.

gmagogsfm commented on torch/nn/functional.py:
```diff
@@ -1509,7 +1509,7 @@ def softmax(input, dim=None, _stacklevel=3, dtype=None):
     return ret
 
 
-def gumbel_softmax(logits, tau=1, hard=False, eps=1e-10, dim=-1):
```
Why do these numbers need to be changed? Is type parsing mistaking them to be ints?

SplitInfinity replied:
Yep.

SplitInfinity commented:
@gmagogsfm  
I don't think this is ready to land so please undo your approval. I just wanted to get your (and @eellison's) attention to discuss the three big issues I identified.

SplitInfinity commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -217,13 +217,81 @@ TypePtr ScriptTypeParser::parseType(const std::string& str) {
   if (default_exprs.empty())
     return default_values;
+
+  // If any default argument type is BroadcastList, the expression
```
This code uses a graph and graph executor to evaluate default arguments.`BroadcastList` is a type that doesn't exist in the JIT type system, so it cannot be used as the output type of a graph. It is meant to be used to convert raw integer arguments into list during compilation. On master, there is code that converts a raw integer to a list if its argument is typed as `BroadcastListX` and attaches that list into the function schema (which has already been created during compilation, which at present ignores default args). But this PR does not ignore default args during compilation, so you can think of these changes as the compile time version of converting an int/float Python object into an IValue list and attaching it onto the function schema.

First, we evaluate the default arguments, and then broadcast later, without a graph or graph executor if needed.

eellison commented on torch/functional.py:
```diff
@@ -1299,6 +1303,23 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
             else:
                 return _VF.norm(input, p, _dim, keepdim=keepdim, dtype=dtype, out=out)  # type: ignore
 
+    if isinstance(p, float):
```
This is a pretty invasive change on our core codebase that makes this less readable - we should sync on other solutions before landing.

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default args in class types"


**Summary**
This commit adds support for default args in methods of TorchScript
classes. Default values are already represented in the JIT TreeView
representation of function (in the `defaultValue` member of the `Param`
subclass of `TreeView`); this commit adds code to parse out the default
values and pack them into the `TreeView` created for the class definition.
This behaviour is gated by a new argument `parse_defaults` that has been
added to several functions in `frontend.py` to make sure that this new
behaviour is triggered only when TorchScript classes are being compiled.
The same code is used to compile functions and interfaces, and default
arguments are not expected to be parsed during compilation of these
entities.

**Test Plan**
This commit adds a unit test to `TestClassType` to test this feature.

`python test/test_jit.py TestClassType.test_default_args`

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

SplitInfinity commented:
Okay, I think it's ready. Major changes since last review:
1. Checks for mutable defaults have been moved to the AST parsing level.
2. Variables cannot be used as default values due to issues with closing over them. A workaround is to initialize the arg to `None` and then initialize with the true default value in the function body.
3. Overload compilation has been migrated to use default values from the AST as well. This required adding code to copy default values from the implementation decl to the overload decl (this allowed me to undo the edits to the `norm` kernel).

SplitInfinity commented:
Ping @eellison and @gmagogsfm 

There are a lot of internal failures but I sampled a handful and they seem to be ones where the argument is annotated as `float` but initialized as `int` (or something similar). They should be easy to fix, but I'd like to deal with those that after all of the common changes look okay so that I don't have to bounce back and forth between fbcode and Github.

eellison commented:
@SplitInfinity sure, i'll take a look today. It's a big PR... If it's not too much difficulty, it's probably worth adding special handling so that we can take an int default for a float argument.

SplitInfinity commented:
`int` default value annotated `float` is easy but there are some other ones like `int` annotated as `<custom type>`

eellison commented:
What's an example of the custom type one ? Yea probably just int as float is fine.

eellison reviewed (commented):
I didn't do an in-depth review yet, wanted to sync up on some aspects of the approach. 

So, as far as I understand, the reason we're switching a lot of the mechanics of how we get and generate default values is because we no longer have access to an object instance of a class, as we did with nn.Modules. (please point out if this is not true). 

I want to point out that this doesn't actually prohibit us from getting the default values: 
```
closed_over_var = 1
class A(object):
    def forward(self, b=closed_over_var, c=4):
        return a  
print(A.forward.__defaults__)
# (1, 4)
```

This also has the nice property that we don't have to regress the scriptability of cases above with `closed_over_var`. The default value is bound at initialization time so there is no (semantic) reason why we couldn't support this case. 

I think using the python value would help with the complexity/brittleness of this stuff
```
      .value("ListLiteralKind", TK_LIST_LITERAL)
      .value("DictLiteralKind", TK_DICT_LITERAL)
      .value("TupleLiteralKind", TK_TUPLE_LITERAL)
      .value("VarKind", TK_VAR)
```

Which atm wouldn't handle stuff like a default value'd enum, or other future things that get added. 

One other question I had was - we do all of the stitching/evaluation of python defaults in the `script_init.cpp` file. Could we add a check for mutable values there ? Since this is only the python compilation path it shouldn't get affected by the broadcasting list defs in native_functions.yaml

I understand this is a difficult PR to land that is a really nice improvement in functionality. We can consider what aspects we consider necessary for fixing and which would just be nice to have, and then decide what to land with after.

SplitInfinity commented:
> I didn't do an in-depth review yet, wanted to sync up on some aspects of the approach.
> 
> So, as far as I understand, the reason we're switching a lot of the mechanics of how we get and generate default values is because we no longer have access to an object instance of a class, as we did with nn.Modules. (please point out if this is not true).
> 
> I want to point out that this doesn't actually prohibit us from getting the default values:
> 
> ```
> closed_over_var = 1
> class A(object):
>     def forward(self, b=closed_over_var, c=4):
>         return a  
> print(A.forward.__defaults__)
> # (1, 4)
> ```
Yeah, I think we could pack all those defaults into a some map and pass it into `jit_script_class_compile` and then reuse `getSchemaWithNameAndDefaults` on all of the methods and properties and their defaults contained in that map. The only disadvantage to this approach is that the default values are not known at compile time because they are added to the schema after compilation. But that's the case with scripted functions anyway. I wish I had thought of that first T_T

> One other question I had was - we do all of the stitching/evaluation of python defaults in the `script_init.cpp` file. Could we add a check for mutable values there ? Since this is only the python compilation path it shouldn't get affected by the broadcasting list defs in native_functions.yaml

I think we could, but we would do the check after compilation. By doing the check in `frontend.py`, we are able to check for that and avoid compilation if any defaults are mutable.

SplitInfinity commented on test/distributed/nn/jit/test_instantiator.py:
```diff
@@ -33,7 +33,7 @@ def test_get_arg_return_types_from_interface(self):
         ) = instantiator.get_arg_return_types_from_interface(MyModuleInterface)
         self.assertEqual(args_str, "tensor, number, word")
-        self.assertEqual(arg_types_str, "tensor: Tensor, number: int, word: str")
+        self.assertEqual(arg_types_str, "tensor: Tensor, number: int, word: str = \"default\"")
```
I think this is some test that checks that a schema is parsed in a specific way and the test was written in a way that didn't include defaults (because they weren't being parsed).

SplitInfinity commented on test/test_jit.py:
```diff
@@ -2302,11 +2305,8 @@ def simple_fn(x, a=a, b=b, c=outer_var + outer_var2):
-
         @torch.jit.script
-        def bool_fn(x, a=outer_c, flag=outer_flag):
+        def bool_fn(x, a: Tensor = torch.tensor(9), flag: Tensor = torch.tensor(False)):
```
This was needed due to issue with closing over variables used as default argument values.

SplitInfinity commented on test/test_jit.py:
```diff
@@ -2676,7 +2673,7 @@ def test_code_with_constants(self):
         """
         @torch.jit.script
-        def foo(x=torch.ones(1)):
+        def foo(x: Tensor = torch.ones(1)):
```
Now that the defaults are being parsed from the AST, there's no way to infer types for the default argument expressions without making a small graph containing them and running it.

eellison commented:
> I think we could, but we would do the check after compilation. By doing the check in frontend.py, we are able to check for that and avoid compilation if any defaults are mutable.

I don't think this is a compelling reason for the extra complexity and maintenance. How often do people use default values that we're optimizing for their failure case? It would be just as easy to iterate over the default values before compilation in the other approach.

> The only disadvantage to this approach is that the default values are not known at compile time because they are added to the schema after compilation. But that's the case with scripted functions anyway. I wish I had thought of that first T_T

Yea, it's true that this PR does add that functionality, which is really nice. It does regress other aspects of scripting though (closing over named constants and int as floats is the big one). I would prefer if we tried to improve that aspect of scripting in a separate PR.

SplitInfinity commented:
I just posted #45098 that addresses the same GitHub issue as this PR but with the same approach that we use now for methods of modules and functions (getting the Python objects for all of the default args, converting them to IValues and attaching them to the schemas of the class methods). Maybe it's better to land that instead of this one?

SplitInfinity commented:
F
